server/src/models/mod.rs

```diff
@@ -61,20 +71,109 @@ impl ErrorBound {
     }
 }
 
-/// Enable equal and not equal for `ErrorBound` and `f32`.
+/// Enable equal and not equal for [`ErrorBound`] and [`f32`].
 impl PartialEq<ErrorBound> for f32 {
     fn eq(&self, other: &ErrorBound) -> bool {
         self.eq(&other.0)
     }
 }
 
-/// Enable less than and greater than for `ErrorBound` and `f32`.
+/// Enable less than and greater than for [`ErrorBound`] and [`f32`].
 impl PartialOrd<ErrorBound> for f32 {
     fn partial_cmp(&self, other: &ErrorBound) -> Option<Ordering> {
         self.partial_cmp(&other.0)
     }
 }
 
+/// Model that uses the fewest number of bytes per value.
+pub struct SelectedModel {
+    /// Id of the model type that created this model.
+    pub model_type_id: u8,
+    /// Index of the last data point in the `UncompressedSegment` that the
+    /// selected model represents.
+    pub end_index: usize,
+    /// The selected model's minimum value.
+    pub min_value: Value,
+    /// The selected model's maximum value.
+    pub max_value: Value,
+    /// Data required in addition to `min` and `max` for the model to
+    /// reconstruct the values it represents when given a specific timestamp.
+    pub values: Vec<u8>,
+}
+
+impl SelectedModel {
+    fn new(
+        model_type_id: u8,
+        end_index: usize,
+        min_value: Value,
+        max_value: Value,
+        values: Vec<u8>,
+    ) -> Self {
+        Self {
+            model_type_id,
+            end_index,
+            min_value,
+            max_value,
+            values,
+        }
+    }
+}
+
+/// Select the model that uses the fewest number of bytes per value.
+pub fn select_model(
+    start_index: usize,
+    pmc_mean: PMCMean,
+    swing: Swing,
+    gorilla: Gorilla,
+    uncompressed_values: &ValueArray,
+) -> SelectedModel {
+    let bytes_per_value = [
+        (PMC_MEAN_ID, pmc_mean.get_bytes_per_value()),
+        (SWING_ID, swing.get_bytes_per_value()),
+        (GORILLA_ID, gorilla.get_bytes_per_value()),
+    ];
+    dbg!(bytes_per_value);
```

Comment:
The documentation for this macro describes it as "quick and dirty debugging". Should this be removed?

Reply:
I noticed that there is a general lack of log messages in the compression process. One option would be to add some log messages throughout the process and then change this `dbg!` to a log message as well.

Alternatively logging could be handled outside the compression process with a log message before calling `try_compress` and then a log message once it returns.

Reply:
Yes it should be removed. I did not add log messages as I assumed a log message would be written by the `StorageEngine` before calling `try_compress` and then again once it returns